### ISSUE TITLE
Option hardware.amdHybridGraphics.disable does not work

### ISSUE THREAD

**Issue by lheckemann** [NixOS/nixpkgs#19775]
## Issue description

The systemd unit generated by [amd-hybrid-graphics.nix](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/hardware/amd-hybrid-graphics.nix) fails to disable the discrete graphics, but does not indicate failure.

Output of `systemctl status amd-hybrid-graphics.service`:

```
● amd-hybrid-graphics.service - Disable AMD Card
   Loaded: loaded (/nix/store/hrf4phk7aqi8yycdcigiv1zk6db42332-unit-amd-hybrid-graphics.service/amd-hybrid-graphics.service; static; vendor preset: enabled)
   Active: active (exited) since Sat 2016-10-22 17:47:21 BST; 25min ago
 Main PID: 800 (code=exited, status=0/SUCCESS)
    Tasks: 0 (limit: 4915)
   CGroup: /system.slice/amd-hybrid-graphics.service

Oct 22 17:47:21 computer systemd[1]: Starting Disable AMD Card...
Oct 22 17:47:21 computer sh[800]: /nix/store/a6gkm5vdylpjnsp9wzcb0c7kz11a3pak-bash-4.3-p46/bin/sh: /sys/kernel/debug/vgaswitcheroo/switch: No such file or directory
Oct 22 17:47:21 computer systemd[1]: Started Disable AMD Card.
```

The success status code is set by `exit 0` in the shell script in the unit file, even though there was clearly a problem with it. The debugfs entry _does_ exist by the time I can access a shell on the system, and I can still disable the discrete graphics manually by running `sudo tee /sys/kernel/debug/vgaswitcheroo/switch <<<IGD`.
### Steps to reproduce

Set `hardware.amdHybridGraphics.disable` to true in the system configuration.
## Technical details
- System: NixOS 17.03pre93673.09e4b78 (Gorilla)
- Nix version: nix-env (Nix) 1.11.4
- Nixpkgs version: "17.03pre93673.09e4b78"
- Hardware: MacbookPro8,2 with dual Intel/AMD graphics chips

**Comment by Timoha** [comment]
Not sure if this is related, but I also have MacbookPro8,2 and when setting `hardware.amdHybridGraphics.disable=true` the screen goes completely black. I've examined `journalctl` of the boot and got these errors that could be useful?

```
kernel: radeon 0000:01:00,0: Invalid ROM contents
kernel: radeon 0000:01:00,0: Invalid ROM contents
```

Also, when I tried manually running `echo DIGD > /sys/kernel/debug/vgaswitcheroo/switch`, it turned to black screen right away. However, setting back with `echo DIGD > /sys/kernel/debug/vgaswitcheroo/switch` it came back to console and `journalctl` didn't show any errors.

- System: NixOS 16.09.1066.417e04f (Flounder)

**Comment by lheckemann** [comment]
Which bootloader do you use, and do you have any other stuff in place to mess with the graphics configuration? I believe that the hybrid graphics not working with GRUB is a known issue which is due to GRUB not doing the EFI boot quite right (using `linux` rather than `linuxefi`), at least on Debian. I haven't understood the full details, but it's something to do with the EFI placing the graphics ROM somewhere in memory and GRUB's `linux` command not handling that right/destroying the mapping or something. I'm not sure what the state is on NixOS, as I'm using systemd-boot. The only way I could get it to boot at all from GRUB was to switch to the integrated graphics before actually loading the kernel, in GRUB itself. https://wiki.archlinux.org/index.php/MacBookPro8,1/8,2/8,3_(2011) has an example of the commands, of course they'd be installed differently on NixOS.

**Comment by lheckemann** [comment]
Adding this to my `configuration.nix` fixed the problem I explained in the initial issue btw. I'd have created a PR to apply the changes to `amd-hybrid-graphics.nix` by now but I'm still somewhat confused with regards to the requirements for testing changes when submitting a PR...
```
  hardware.amdHybridGraphics.disable = true;
  systemd.services."amd-hybrid-graphics".requires = ["vgaswitcheroo.path"];
  systemd.services."amd-hybrid-graphics".wantedBy = lib.mkForce [];
  systemd.services."amd-hybrid-graphics".before = ["systemd-vconsole-setup.service" "display-manager.service"];
  systemd.paths."vgaswitcheroo".pathConfig.PathExists = "/sys/kernel/debug/vgaswitcheroo/switch";
  systemd.paths."vgaswitcheroo".pathConfig.Unit = "amd-hybrid-graphics.service";
  systemd.paths."vgaswitcheroo".wantedBy = ["multi-user.target"];
```

**Comment by Timoha** [comment]
@lheckemann yes, I'm also using systemd-boot. I don't have any other config for the graphics, in fact, I have the default configuration since I only started playing with nixos. Never gotten it to boot into GUI though.

I give GRUB a try since I used it before to boot Ubuntu and all the hacks to disable radeon card worked well.

**Comment by lheckemann** [comment]
Yeah, GRUB with the `outb` hacks is probably the easier option. It took me a lot of fiddling to get where I am with it, and it doesn't actually provide any tangible benefits over GRUB as far as I can tell.

**Comment by lheckemann** [comment]
Closed by #21227.
